eth/state_accessor.go

```diff
@@ -239,12 +240,37 @@
 	if err != nil {
 		return nil, vm.BlockContext{}, nil, nil, err
 	}
+	// upgrade build-in system contract before normal txs if Feynman is not enabled
+	if !eth.blockchain.Config().IsFeynman(block.Number(), block.Time()) {
+		systemcontracts.UpgradeBuildInSystemContract(eth.blockchain.Config(), block.Number(), parent.Time(), block.Time(), statedb)
+	}
 	if txIndex == 0 && len(block.Transactions()) == 0 {
 		return nil, vm.BlockContext{}, statedb, release, nil
 	}
 	// Recompute transactions up to the target index.
-	signer := types.MakeSigner(eth.blockchain.Config(), block.Number(), block.Time())
+	var (
+		signer         = types.MakeSigner(eth.blockchain.Config(), block.Number(), block.Time())
+		beforeSystemTx = true
+	)
 	for idx, tx := range block.Transactions() {
+		// upgrade build-in system contract before system txs if Feynman is enabled
+		if beforeSystemTx {
```

Comment:
Is it ok to move this `posa, ok := eth.Engine().(consensus.PoSA)` before the loop?